packages/plugin-restify/src/restify.js

```diff
@@ -0,0 +1,74 @@
+const domain = require('domain') // eslint-disable-line
+const extractRequestInfo = require('./request-info')
+const createReportFromErr = require('@bugsnag/core/lib/report-from-error')
+const handledState = {
+  severity: 'error',
+  unhandled: true,
+  severityReason: {
+    type: 'unhandledErrorMiddleware',
+    attributes: { framework: 'Restify' }
+  }
+}
+
+module.exports = {
+  name: 'restify',
+  init: client => {
+    const requestHandler = (req, res, next) => {
+      const dom = domain.create()
+
+      // Start a session whether sessions are used or not. We use this
+      // to store request-specific info, in case of any errors.
+      const sessionClient = client.startSession()
+
+      // attach it to the request
+      req.bugsnag = sessionClient
+
+      // extract request info and pass it to the relevant bugsnag properties
+      const requestInfo = extractRequestInfo(req)
+      sessionClient.metaData = { ...sessionClient.metaData, request: requestInfo }
+      sessionClient.request = {
+        clientIp: requestInfo.clientIp,
+        headers: requestInfo.headers,
+        httpMethod: requestInfo.httpMethod,
+        url: requestInfo.url,
+        referer: requestInfo.referer
+      }
+
+      // unhandled errors caused by this request
+      dom.on('error', (err) => {
+        req.bugsnag.notify(createReportFromErr(err, handledState), {}, (e, report) => {
+          if (e) return client._logger('Failed to send report to Bugsnag')
+          req.bugsnag.config.onUncaughtException(err, report, client._logger)
+        })
+        if (!res.headersSent) {
+          const body = 'Internal server error'
+          res.writeHead(200, {
+            'Content-Length': Buffer.byteLength(body),
+            'Content-Type': 'text/plain'
+          })
+          res.end(body)
+        }
+      })
+
+      return dom.run(next)
+    }
+
+    const errorHandler = (req, res, err, cb) => {
+      if (err.statusCode && err.statusCode < 500) return cb()
+      let c = req.bugsnag
+      if (!c) {
+        client._logger.warn(
+          'req.bugsnag is not defined. Make sure the @bugsnag/plugin-restify requestHandler middleware is added first.',
+          'Some request information will be missing.'
+        )
+        c = client
+      }
+      c.notify(createReportFromErr(err, handledState))
```

Comment:
in the Koa integration we attach the req information when we are using a client that was not on the request which seems like a better pattern than logging a warning that we couldn't add the request information. Is that possible here?

Reply:
It's possible. There's certainly inconsistency here – I this approach for restify and express but not koa.

My intention was for it to be clear that omiting the `requestHandler` is a usage error. When you don't add the `requestHandler` the following bad things happen:

- there's no `req.bugsnag()` for handled errors with request context
- there's no session tracking

It's possible to make all three web frameworks take the same approach, I think we just need to decide. My suggestion is the following (y not both!!!):

```js
const requestHandler = () => {
  // start a session
  // capture request info
  // add sesstion to req/ctx.bugsnag
}

const errorHandler = () => {
  // check for req/ctx.bugsnag
  // if it exists use its notify function
  // if it doesn't exist, warn
  // collect request info
  // notify with request info
}
```

Reply:
> It's possible to make all three web frameworks take the same approach, I think we just need to decide. My suggestion is the following (y not both!!!):

Consistency here seems ideal. Not using the `requestHandler` should log a warning, but following the warning we should collect as much as possible from the available environment and notify.